proposals/003-authentication-api.md

```diff
@@ -0,0 +1,223 @@
+      returned by the broker. This is a technique that each identity-dependent filter in the chain could use, but each needs to reimplement the logic.
+    - For "SASL termination", a `Filter` implementation could handle the `SaslAuthenticateRequest` frames itself 
+        - If it makes its own `SaslAuthenticateRequests` (with different credentials, and/or a different mechanism) to the server then subsequent filters in the chain need to use the "SASL passthrough" technique to be aware of the broker-facing identity, and there's no way for them to know the original identity.
+        - If it doesn't (so that there is no `SaslAuthenticate` intersection with the broker) then subsequent filters in the chain have no way of knowing the client's identity.
+
+## Motivation
+
+The lack of this API makes implementing client identity aware plugins difficult, or impossible.
+
+Goals: 
+
+* Enable `Filters` to access a client's identity using a single, consistent API, irrespective of what authentication mechanism is implemented, TLS or SASL, and whether it's implemented by the proxy runtime (in the TLS case), or a prior Filter in the chain (in the SASL termination case).
+* Don't require a `Filter` to handle `SaslAuthenticate` unless it is performing SASL termination. 
+
+Non-goals:
+
+* Defining an API for exposing the identity of a _broker_ to `Filters` (in cases where the proxy mututally authenticates the broker).
+
+## Proposal
+
+### Public API changes
+
+Add the following interface to the `kroxylicious-api` module to allow a `Filter` implementation to opt into consuming authentication outcomes:
+
+```java
+package io.kroxylicious.proxy.filter;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+// This is a new interface in the api module
+interface ClientSubjectAwareFilter extends Filter {
+    // Called when a client authenticates, or reauthenticates
+    void onClientAuthentication(Subject clientSubject, FilterContext context);
+    // Called when a client fails authentication, or reauthentication
+    default void onClientAuthenticationFailure(LoginException exception, FilterContext context) { 
+    }
+}
+```
+
+Add the following two methods to `FilterContext` to allow SASL terminating `Filter` implementations to propagate their authentication outcomes to upstream filters:
+
+```java
+public interface FilterContext {
+
+    void clientAuthenticationSuccess(Subject subject);
+
+    void clientAuthenticationFailure(LoginException e);
+    
+}
+```
+
+### Runtime changes
+
+1. Change the `kroxylicious-runtime` module so a `Subject` gets instantiated on initial connection.
+2. On successful completion of a TLS handshake, associate a `javax.security.auth.x500.X500Principal` with that `Subject`.
+3. On construction of the filter chain fire an internal message down the pipeline.
```

Comment:
The goal of the proposal said:

> Enable `Filters` to access a client's identity using a single, consistent API, irrespective of what authentication mechanism is implemented, TLS or SASL,

I see how point 2 satisfies the TLS identity.  What catching the SASL identity and adding a subject for it?

Reply:
Yes, you'd need a plugin implementation to do SASL sniffing. But it wouldn't be in the runtime.

Reply:
It would be nice if we had a mechanism at proxy start up to validate the non-emptiness of intersection of set of principal types emitted by plugins and the set of principal types required by `ClientSubjectAware` plugins.

Reply:
ah so the user says I'm interested in SASL identities, so I'll install a SASL sniffer early on in the chain?

Reply:
closing this as outdated by the latest changes